From caac9e918fc59c8dfe5c820a3d7a43b948112550 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Horia=20Geant=C4=83?= <horia.geanta@nxp.com>
Date: Wed, 20 Jan 2016 12:16:18 +0200
Subject: [PATCH 168/752] crypto: caam - add support for PKC keygen
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

This is a squash of the following commits:
fc281aa59e4c6f2e1574d4a6812438a6e06be0a9 "added support for PKC keygen"
7191bc627c7038611663f90c7dd053733bc5586b "crypto: caampkc - workaround for erratum A-006899"
followed by merge conflicts resolution.

"As a part of PKC support, RSA, DSA DH, ECDH, ECDSA requires key
generation. The patch adds support for key generation support
for DSA, ECDSA, DH, ECDH.

The patch adds DH operation support too."

Signed-off-by: Yashpal Dutta <yashpal.dutta@freescale.com>
Signed-off-by: Tudor Ambarus <tudor-dan.ambarus@nxp.com>
Signed-off-by: Horia GeantÄƒ <horia.geanta@nxp.com>
[Original patch taken from QorIQ-SDK-V2.0-20160527-yocto]
Signed-off-by: Yanjiang Jin <yanjiang.jin@windriver.com>
---
 drivers/crypto/caam/pkc_desc.h | 24 ++++++++++++++++++++++++
 1 file changed, 24 insertions(+)

diff --git a/drivers/crypto/caam/pkc_desc.h b/drivers/crypto/caam/pkc_desc.h
index 56fc2ee..e859e40 100644
--- a/drivers/crypto/caam/pkc_desc.h
+++ b/drivers/crypto/caam/pkc_desc.h
@@ -176,9 +176,11 @@ struct rsa_edesc {
  */
 struct dsa_edesc_s {
 	enum pkc_req_type req_type;
+	enum curve_t curve_type;
 	uint32_t l_len;
 	uint32_t n_len;
 	dma_addr_t key_dma;
+	dma_addr_t s_dma;
 	dma_addr_t f_dma;
 	dma_addr_t q_dma;
 	dma_addr_t r_dma;
@@ -186,8 +188,28 @@ struct dsa_edesc_s {
 	dma_addr_t d_dma;
 	dma_addr_t ab_dma;
 	dma_addr_t g_dma;
+	dma_addr_t g_sg_dma;
 	dma_addr_t tmp_dma;
 	uint8_t *tmp; /* Allocate locally for dsa_verify */
+	struct sec4_sg_entry g_sg;
+	bool erratum_A_006899;
+	u32 hw_desc[];
+};
+
+/*
+ * dh_edesc - s/w-extended for dh and ecdh descriptors
+ * @hw_desc: the h/w job descriptor
+ */
+struct dh_edesc_s {
+	enum pkc_req_type req_type;
+	enum curve_t curve_type;
+	uint32_t l_len;
+	uint32_t n_len;
+	dma_addr_t q_dma;
+	dma_addr_t ab_dma;
+	dma_addr_t w_dma;
+	dma_addr_t s_dma;
+	dma_addr_t z_dma;
 	u32 hw_desc[];
 };
 
@@ -197,5 +219,7 @@ void *caam_rsa_priv_f2_desc(struct rsa_edesc *);
 void *caam_rsa_priv_f3_desc(struct rsa_edesc *);
 void *caam_dsa_sign_desc(struct dsa_edesc_s *);
 void *caam_dsa_verify_desc(struct dsa_edesc_s *);
+void *caam_keygen_desc(struct dsa_edesc_s *);
+void *caam_dh_key_desc(struct dh_edesc_s *);
 
 #endif
-- 
2.9.3

